File: backend/app/schemas/social.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

PostType = Literal["photo", "video", "progress", "achievement", "workout", "run"]
# 'private' renomeado pra 'followers' — ver comentario em models/social.py.
PostVisibility = Literal["public", "followers"]


class PostCreate(BaseModel):
    type: PostType
    caption: str | None = None
    media_url: str | None = None
    visibility: PostVisibility = "public"
    reference_id: uuid.UUID | None = None


class PostOut(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    author: str
    type: str
    caption: str | None = None
    media_url: str | None = None
    visibility: str
    reference_id: uuid.UUID | None = None
    created_at: datetime
    likes_count: int
    comments_count: int
    is_liked_by_me: bool = False


class FollowOut(BaseModel):
    id: uuid.UUID
    follower_id: uuid.UUID
    following_id: uuid.UUID
    created_at: datetime

    class Config:
        from_attributes = True


class LikeOut(BaseModel):
    id: uuid.UUID
    post_id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime

    class Config:
        from_attributes = True


class CommentCreate(BaseModel):
    content: str


class CommentOut(BaseModel):
    id: uuid.UUID
    post_id: uuid.UUID
    user_id: uuid.UUID
    author: str
    content: str
    created_at: datetime


class UserBrief(BaseModel):
    id: uuid.UUID
    name: str

File: mobile/services/social.ts
import { api } from '@/services/api';

export type PostType = 'photo' | 'video' | 'progress' | 'achievement' | 'workout' | 'run';
export type PostVisibility = 'public' | 'private';

export interface Post {
  id: string;
  user_id: string;
  author: string;
  type: string;
  caption: string | null;
  media_url: string | null;
  visibility: string;
  reference_id: string | null;
  created_at: string;
  likes_count: number;
  comments_count: number;
  is_liked_by_me: boolean;
}

export interface PostCreatePayload {
  type: PostType;
  caption?: string | null;
  media_url?: string | null;
  visibility?: PostVisibility;
}

export interface Comment {
  id: string;
  post_id: string;
  user_id: string;
  author: string;
  content: string;
  created_at: string;
}

export interface UserBrief {
  id: string;
  name: string;
}

export async function getFeed(limit: number, offset: number): Promise<Post[]> {
  const response = await api.get<Post[]>('/feed', { params: { limit, offset } });
  return response.data;
}

export async function createPost(payload: PostCreatePayload): Promise<Post> {
  const response = await api.post<Post>('/posts', payload);
  return response.data;
}

export async function getPost(postId: string): Promise<Post> {
  const response = await api.get<Post>(`/posts/${postId}`);
  return response.data;
}

export async function deletePost(postId: string): Promise<void> {
  await api.delete(`/posts/${postId}`);
}

export async function listUserPosts(userId: string): Promise<Post[]> {
  const response = await api.get<Post[]>(`/users/${userId}/posts`);
  return response.data;
}

export async function followUser(userId: string): Promise<void> {
  await api.post(`/users/${userId}/follow`);
}

export async function unfollowUser(userId: string): Promise<void> {
  await api.delete(`/users/${userId}/follow`);
}

export async function listFollowers(userId: string): Promise<UserBrief[]> {
  const response = await api.get<UserBrief[]>(`/users/${userId}/followers`);
  return response.data;
}

export async function listFollowing(userId: string): Promise<UserBrief[]> {
  const response = await api.get<UserBrief[]>(`/users/${userId}/following`);
  return response.data;
}

export async function likePost(postId: string): Promise<void> {
  await api.post(`/posts/${postId}/like`);
}

export async function unlikePost(postId: string): Promise<void> {
  await api.delete(`/posts/${postId}/like`);
}

export async function listComments(postId: string): Promise<Comment[]> {
  const response = await api.get<Comment[]>(`/posts/${postId}/comments`);
  return response.data;
}

export async function createComment(postId: string, content: string): Promise<Comment> {
  const response = await api.post<Comment>(`/posts/${postId}/comments`, { content });
  return response.data;
}

export async function deleteComment(commentId: string): Promise<void> {
  await api.delete(`/comments/${commentId}`);
}

/**
 * O backend serializa datetimes "ingenuos" em UTC sem sufixo de fuso —
 * sem tratar isso, o JS interpretaria a string como horario local.
 */
export function parseUtcDate(isoDate: string): Date {
  const hasTimezone = /[zZ]|[+-]\d{2}:?\d{2}$/.test(isoDate);
  return new Date(hasTimezone ? isoDate : `${isoDate}Z`);
}

export function formatPostDate(isoDate: string): string {
  const date = parseUtcDate(isoDate);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffMinutes = Math.floor(diffMs / 60000);
  if (diffMinutes < 1) return 'agora';
  if (diffMinutes < 60) return `${diffMinutes}min`;
  const diffHours = Math.floor(diffMinutes / 60);
  if (diffHours < 24) return `${diffHours}h`;
  const diffDays = Math.floor(diffHours / 24);
  if (diffDays < 7) return `${diffDays}d`;
  return date.toLocaleDateString('pt-BR', { day: '2-digit', month: 'short' });
}
